docs/tools.md

```diff
@@ -720,15 +723,556 @@ def my_flaky_tool(query: str) -> str:
+
+To add or remove available toolsets, you can wrap a [`CombinedToolset`](#combining-toolsets) and replace it during the run with one that can include fewer, more, or entirely different toolsets.
+
+```python {title="wrapper_toolset.py" requires="function_toolset.py"}
+from function_toolset import weather_toolset, datetime_toolset
+
+from pydantic_ai import Agent, RunContext
+from pydantic_ai.models.test import TestModel
+from pydantic_ai.toolsets import WrapperToolset, FunctionToolset
+
+togglable_toolset = WrapperToolset(weather_toolset)
+
+def toggle(ctx: RunContext[WrapperToolset]):
+    if ctx.deps.wrapped == weather_toolset:
+        ctx.deps.wrapped = datetime_toolset
+    else:
+        ctx.deps.wrapped = weather_toolset
+
+test_model = TestModel() # (1)!
+agent = Agent(
+    test_model,
+    deps_type=WrapperToolset, # (2)!
+    toolsets=[togglable_toolset, FunctionToolset([toggle])]
+)
+print([t.name for ts in agent.toolsets for t in ts.tool_defs])
+#> ['temperature_celsius', 'temperature_fahrenheit', 'conditions', 'toggle']
+
+result = agent.run_sync('Toggle the toolset', deps=togglable_toolset)
+print([t.name for t in test_model.last_model_request_parameters.function_tools])
+#> ['now', 'toggle']
+```
+
+1. We're using [`TestModel`][pydantic_ai.models.test.TestModel] here because it makes it easy to see which tools were available on each run.
+2. We're using the agent's dependencies to give the `toggle` tool access to the `togglable_toolset` via the `RunContext` argument
+
+### Building a Custom Toolset
+
+To define a fully custom toolset with its own logic to list available tools and handle them being called, you can subclass [`BaseToolset`][pydantic_ai.toolsets.BaseToolset] or [`AsyncBaseToolset`][pydantic_ai.toolsets.AsyncBaseToolset], depending on whether listing the available tools can be done synchronously or requires an asynchronous network request. Tool calls themselves are always implemented asynchronously.
+
+These abstract classes require you to implement [`tool_defs()`][pydantic_ai.toolsets.AbstractToolset.tool_defs] or [`async_tool_defs()`][pydantic_ai.toolsets.AsyncBaseToolset.async_tool_defs] respectively, as well as the [`_max_retries_for_tool()`][pydantic_ai.toolsets.AbstractToolset._max_retries_for_tool], [`_get_tool_args_validator()`][pydantic_ai.toolsets.AbstractToolset._get_tool_args_validator] and [`_call_tool()`][pydantic_ai.toolsets.AbstractToolset._call_tool] methods.
+
+If you want to reuse a network connection or session across tool listings and calls during an agent run step, you can implement [`__aenter__()`][pydantic_ai.toolsets.AbstractToolset.__aenter__] and [`__aexit__()`][pydantic_ai.toolsets.AbstractToolset.__aexit__], which will be called when the agent that uses the toolset is itself entered using the [`async with agent`][pydantic_ai.Agent.__aenter__] context manager.
```

Comment:
Note we'll need to do this differently with temporal, but that's fine. Just this section might get some tweaking..